hyper/http11/response.py

```diff
@@ -48,7 +49,8 @@ def __init__(self, code, reason, headers, sock, connection=None):
         # bother checking for content-length, we just keep reading until
         # we no longer can.
         self._expect_close = False
-        if b'close' in self.headers.get(b'connection', []):
+        if (b'close' in self.headers.get(b'connection', []) or
+                b'close' in self.headers.get(b'proxy-connection', [])):
             self._expect_close = True
```

Comment:
In the response above a proxy server exclicitly set that this is a HTTP/1.0 response (not HTTP/1.1), didn't set a `Content-Length` header and closed connection after sending body. This is a valid HTTP/1.0 response. 

[RFC 1945, HTTP/1.0, section 7.2.2 Length](https://tools.ietf.org/search/rfc1945#section-7.2.2):
> When an Entity-Body is included with a message, the length of that
> body may be determined in one of two ways. If a Content-Length header
> field is present, its value in bytes represents the length of the
> Entity-Body. Otherwise, the body length is determined by the closing
> of the connection by the server.

[RFC 1945, HTTP/1.0, section 1.3 Overall Operation](https://tools.ietf.org/search/rfc1945#section-1.3):
> Except for experimental applications, current practice requires that
> the connection be established by the client prior to each request and
> closed by the server after sending the response.

But it is an invalid response from HTTP/1.1 perspective as you have already fairly mentioned.

[RFC 7230, HTTP/1.1 Message Syntax and Routing, section 3.4 Handling Incomplete Messages](https://tools.ietf.org/html/rfc7230#section-3.4):

> [...] A response that has neither chunked
> transfer coding nor Content-Length is terminated by closure of the
> connection and, thus, is considered complete regardless of the number
> of message body octets received, provided that the header section was
> received intact.

[RFC 2068, HTTP/1.1, section 14.10 Connection](https://tools.ietf.org/html/rfc2068#section-14.10):
> HTTP/1.1 applications that do not support persistent connections MUST
> include the "close" connection option in every message.
> 
> In HTTP/1.0, each connection is established by the client prior to
> the request and closed by the server after sending the response.
> However, some implementations implement the explicitly negotiated
> ("Keep-Alive") version of persistent connections described in Section
> 19.7.1 of [RFC2068].

Also it is expected, but not strictly required, for HTTP/1.1 clients to understand valid HTTP/1.0 responses as well (as per [RFC 2068, HTTP/1.1, section 19.7 Compatibility with Previous Versions](https://tools.ietf.org/html/rfc2068#section-19.7)).

So the key difference between HTTP/1.1 and HTTP/1.0 responses for that matter, as I see it, is that HTTP/1.1 response must include a `connection: close` header if a server is about to close connection in the absence of `content-length`/chunked encoding headers, but for HTTP/1.0 response connection should always be closed anyway (except when `connection: keep-alive` is present, see below).

I suggest to set `self._expect_close` to True for HTTP/1.0 responses which doesn't contain `connection` header.

[RFC 2068, HTTP/1.1, section 19.7.1 Compatibility with HTTP/1.0 Persistent Connections](https://tools.ietf.org/html/rfc2068#section-19.7.1):

> Some clients and servers may wish to be compatible with some previous
> implementations of persistent connections in HTTP/1.0 clients and
> servers. Persistent connections in HTTP/1.0 must be explicitly
> negotiated as they are not the default behavior. [...]
> 
> The following describes the original HTTP/1.0 form of persistent
> connections.
> 
> When it connects to an origin server, an HTTP client MAY send the
> Keep-Alive connection-token in addition to the Persist connection-
> token:
> 
>       Connection: Keep-Alive
> 
> An HTTP/1.0 server would then respond with the Keep-Alive connection
> token and the client may proceed with an HTTP/1.0 (or Keep-Alive)
> persistent connection.

Thus the assertion will not be triggered even if we ignore the `proxy-conenction` header.

An instant question if you consider this proposal acceptable: should this be done in this PR, or in a different one? This change is not just proxy-related and it definitely requires adding some new tests.